### PR TITLE
Use PatchStatus to update node status in kubelet.

### PR DESCRIPTION
**Pull request by Random-Liu**
Fixes https://github.com/kubernetes/kubernetes/issues/37771.

This PR changes kubelet to update node status with `PatchStatus`.

@caesarxuchao @ymqytw told me that there is a limitation of current `CreateTwoWayMergePatch`, it doesn't support primitive type slice which uses strategic merge.
* I checked the node status, the only primitive type slices in NodeStatus are as follows, they are not using strategic merge:
  * [`ContainerImage.Names`](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/v1/types.go#L2963)
  * [`VolumesInUse`](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/v1/types.go#L2909)
* Volume package is already [using `CreateStrategicMergePath` to generate node status update patch](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/attachdetach/statusupdater/node_status_updater.go#L111), and till now everything is fine. 

@yujuhong @dchen1107 
/cc @kubernetes/sig-node

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37871)

### CODE REVIEW

**Comment by k8s-oncall** [comment]
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37871)

**Comment by ymqytw** [comment]
@Random-Liu PATCH doesn't work for MERGING list of primitives, such as [finalizers](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/v1/types.go#L205). Merging list should specify `patchStrategy:"merge"`. Other list of primitives use replace patchStrategy for default.

>ContainerImage.Names
VolumesInUse

The 2 you mentioned above are using replace patchStrategy. I believe PATCH should work for them by replacing the whole list.

**Comment by Random-Liu** [comment]
@ymqytw Thanks for confirmation! :)

**Comment by Random-Liu** [comment]
Will update the PR to fix the unit test.

**Comment by Random-Liu** [comment]
@yujuhong Fixed the unit test. Please take a look. :)

**Comment by rmmh** [comment]
@k8s-bot gci gce e2e test this (job was accidentally deleted)

**Comment by yujuhong** [comment]
@caesarxuchao, could you review the status update operation? Thanks!

File: pkg/util/node/node.go [hunk 1]
```diff
@@ -153,3 +154,27 @@ func SetNodeCondition(c clientset.Interface, node types.NodeName, condition v1.N
...
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal new node %#v for node %q: %v", newNode, nodeName, err)
+	}
+
+	patchBytes, err := strategicpatch.CreateTwoWayMergePatch(oldData, newData, v1.Node{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create patch for node %q: %v", nodeName, err)
+	}
+
+	updatedNode, err := c.Core().Nodes().PatchStatus(string(nodeName), patchBytes)
```
Comment (caesarxuchao) [inline pkg/util/node/node.go]: Could you switch to use the regular Patch() method? It supports specifying subresources now. We should deprecate the special PatchStatus() method.

Comment (Random-Liu) [inline pkg/util/node/node.go]: So now `Patch` can update status, too?

File: pkg/util/node/node.go [hunk 2]
```diff
@@ -153,3 +154,27 @@ func SetNodeCondition(c clientset.Interface, node types.NodeName, condition v1.N
 	_, err = c.Core().Nodes().PatchStatus(string(node), patch)
 	return err
 }
+
+// UpdateNodeStatus updates node status with PatchStatus.
+func UpdateNodeStatus(c clientset.Interface, nodeName types.NodeName, oldNode *v1.Node, newNode *v1.Node) (*v1.Node, error) {
```
Comment (Random-Liu) [inline pkg/util/node/node.go]: Ha, I see. Just set the old one to new one. I see.

File: pkg/util/node/node.go (hunk 1 above)
Comment (caesarxuchao) [inline pkg/util/node/node.go]: Yeah, you need to set the `subreources`.
```
Patch(name string, pt api.PatchType, data []byte, subresources ...string) (result *v1.Node, err error)
```

Comment (Random-Liu) [inline pkg/util/node/node.go]: Done

File: pkg/util/node/node.go (hunk 2 above)
Comment (Random-Liu) [inline pkg/util/node/node.go]: Done

**Comment by Random-Liu** [comment]
The unit test `TestTryRegisterWithApiServer` failed because we are actually using `UpdateStatus` to update `ObjectMeta`(`Annotation`). (See https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_node_status.go#L143-L171)

We could:
1) Explicitly change the function from `UpdateNodeStatus` to `UpdateNode`.
2) Still name the function as `UpdateNodeStatus` and do not reset `ObjectMeta` inside the container.

@caesarxuchao Which do you prefer?

**Comment by Random-Liu** [comment]
Offline discussed with @caesarxuchao, only reset `Node.Spec` in `UpdateNodeStatus`, because:
1. Apiserver allows updating objectmeta when updating node status.
2. Some component does use this to update node annotations.

**Comment by Random-Liu** [comment]
@caesarxuchao @yujuhong Addressed comments. Let's merge this PR after https://github.com/kubernetes/kubernetes/pull/38375 to avoid conflict during cherry-pick.

**Comment by Random-Liu** [comment]
@k8s-bot gce etcd3 e2e test this

**Comment by Random-Liu** [comment]
@caesarxuchao PR ready to go.

**Pull request merged**